File: evolve/kernel/kernel.py
"""
Python wrapper for the window.EvolveKernel JS object (Pyodide environment).

Usage:
    from evolve.kernel import kernel

    node_id = kernel.dom.create("div", {"textContent": "hello"})
    kernel.log("info", "created node %s" % node_id)

    # Register a python callback:
    def on_click(ev):
        print("clicked", ev)
    cb_id = kernel.register_callback(on_click)
    kernel.dom.update(node_id, {"onclick": str(cb_id)})

Notes:
- This module expects Pyodide (pyodide.create_proxy) and `window.EvolveKernel` to exist.
- Keep references to callbacks if you plan to unregister later (this module does that for you).
"""

from typing import Any, Callable
from js import EvolveKernel
from pyodide import create_proxy
import asyncio

# string callback_proxies and pyfunction in dict , so Garbage collector wont remove them
_callback_proxies: dict[int, Any] = {}
# reverse mapping fromm callback ID-->callback Function
_callback_pyfuncs: dict[int, Callable] = {}


def _to_py(js_value: Any) -> Any:
    """
    Converts JS --> Python for objects that support to_py()
    """
    try:
        if hasattr(js_value, "to_py"):
            return js_value.to_py()
    except Exception:
        pass
    return js_value


# LOGGING


def log(level: str, msg: str) -> dict[str, Any]:
    try:
        res = EvolveKernel.log(level, msg)
        return _to_py(res)
    except Exception as e:
        return {"ok": False, "error": str(e)}


# DOM WRAPPERS


class _Dom:
    @staticmethod
    def create(
        tag: str, props: dict[str, Any] | None = None, children: list[Any] | None = None) -> dict[str, Any]:
        props = props or {}
        children = children or []

        try:
            res = EvolveKernel.dom.create(tag, props, children)
            return _to_py(res)
        except Exception as e:
            return {"ok": False, "error": str(e)}
        
    @staticmethod
    def remove(node_id: int) -> dict[str, Any]:
        try:
            res = EvolveKernel.dom.remove(int(node_id))
            return _to_py(res)
        except Exception as e:
            return {"ok": False, "error": str(e)}

    @staticmethod
    def update(nodeID: int, props: dict[str, Any] | None = None) -> dict[str, Any]:
        props = props or {}
        try:
            res = EvolveKernel.dom.update(nodeID, props)
            return _to_py(res)
        except Exception as e:
            return {"ok": False, "error": str(e)}

    @staticmethod
    def append(parentID: int, nodeID: int) -> dict[str, Any]:
        try:
            res = EvolveKernel.dom.append(parentID, nodeID)
            return _to_py(res)
        except Exception as e:
            return {"ok": False, "error": str(e)}

    @staticmethod
    def query(selector: str) -> dict[str, Any]:
        try:
            res = EvolveKernel.dom.query(selector)

            return _to_py(res)
        except Exception as e:
            return {"ok": False, "error": str(e)}


# exposing public interface as kernel.dom.method instead of kernel._DOM.method
# just making it look good.
dom = _Dom()


# FILE SYSTEM WRAPPERS


class _FS:
    
    @staticmethod
    def read(path: str) -> dict[str, Any]:
        try:
            res = EvolveKernel.fs.read(path)
            return _to_py(res)
        except Exception as e:
            return {"ok": False, "error": str(e)}

    @staticmethod
    def write(path: str, contents: Any) -> dict[str, Any]:
        try:
            res = EvolveKernel.fs.write(path, contents)
            return _to_py(res)
        except Exception as e:
            return {"ok": False, "error": str(e)}


fs = _FS()


# NETWORK WRAPPER (ASYNC)


class _NET:
    @staticmethod
    async def fetch(url: str, options: dict[str,Any] | None = None) -> dict[str, Any]:
        options = options or {}

        try:
            res = await EvolveKernel.net.fetch(url, options)
            return _to_py(res)
        except Exception as e:
            return {"ok": False, "error": str(e)}


net = _NET()


# CALLBACK REGISTRATION


def register_callback(py_fun: Callable) -> int:
    """
    Registers python function as JS-callable callback
    """

    if not callable(py_fun):
        raise TypeError("register_callback expects a callable function")

    # create proxy for JS to call python functions
    # *a- accept Any number of arguments
    # **k- accept Any number of keyword arguments
    # then call _call_python_callback function with these arguments
    proxy = create_proxy(lambda *a, **k: _call_python_callback(py_fun, a, k))

    try:
        res = EvolveKernel.registerCallback(proxy)
        res_py = _to_py(res)
        if not res_py.get("ok", False):
            proxy.destroy()
            raise RuntimeError(f"registerCallback failed: {res_py.get('error')}")

        cb_id = int(res_py["value"])
        _callback_proxies[cb_id] = proxy
        _callback_pyfuncs[cb_id] = py_fun
        return cb_id
    except Exception as e:
        try:
            proxy.destroy()
        except Exception:
            pass
        raise e


def _call_python_callback(py_fun: Callable, args: tuple, kargs: dict) -> Any:
    """
    JS --> Python callaback handlers
    """

    py_args = tuple(_to_py(a) for a in args)
    py_kargs = {k: _to_py(v) for k, v in kargs.items()}

    try:
        result = py_fun(*py_args, **py_kargs)
        if asyncio.iscoroutine(result):
            asyncio.ensure_future(result)
        return result
    except Exception as e:
        log("error", f"callback error:{e}")
        return {"ok": False, "error": str(e)}


def unregister_callback(cb_id: int) -> dict[str, Any]:
    cb_id = int(cb_id)

    proxy = _callback_proxies.pop(cb_id, None)
    _callback_pyfuncs.pop(cb_id, None)

    if proxy:
        try:
            proxy.destroy()

        except Exception:
            pass

    try:
        res = EvolveKernel.unregisterCallback(cb_id)
        return _to_py(res)
    except Exception as e:
        return {"ok": False, "error": str(e)}


# KERNEL FACADE
# Public interface for accessibility


class KernelFacade:
    dom = dom
    fs = fs
    net = net
    log = staticmethod(log)
    register_callback = staticmethod(register_callback)
    unregister_callback = staticmethod(unregister_callback)


kernel = KernelFacade()

File: evolve/kernel/kernel.js

// Minimal Evolve Kernel - exposes APIs to WASM / Python.
// Attached to window.EvolveKernel

const EvolveKernel = (function () {
  // internal node registry (maps nodeId -> DOM node)
  const nodes = new Map();
  let nextNodeId = 1;

  // callback registry for WASM/Python to register functions
  const callbacks = new Map();
  let nextCallbackId = 1;

  function genNodeId() {
    return nextNodeId++;
  }
  function genCallbackId() {
    return nextCallbackId++;
  }

  // Logger
  function log(level, message) {
    const payload = { ts: Date.now(), level, message };
    console[level === "error" ? "error" : "log"]("[Kernel]", payload);
    return { ok: true };
  }

  // DOM helpers
  // create dom nodes with specific tags, props and children
  function create(tag, props = {}, children = []) {
    try {
      const el = document.createElement(tag);
      applyProps(el, props);
      const id = genNodeId();
      nodes.set(id, el);
      el.__evolve_id = id
      // attach children if any (children can be nodeIds or strings)
      for (const child of children) {
        if (typeof child === "number") {
          const childNode = nodes.get(child);
          if (childNode) el.appendChild(childNode);
        } else {
          el.appendChild(document.createTextNode(String(child)));
        }
      }
      return { ok: true, value: id };
    } catch (e) {
      return { ok: false, error: e.message };
    }
  }

  function remove(nodeID) {
    try {
      const node = nodes.get(nodeID)

      if (!node) {
        return { ok: false, error: "Node not found" }
      }

      // Recursively remove children first (using DOM tree)
      const children = Array.from(node.children || [])
      for (const child of children) {
        const childId = findNodeId(child)
        if (childId) {
          remove(childId)
        } else {
          child.remove() // orphan case
        }
      }

      // Remove from real DOM
      if (node.parentNode) {
        node.parentNode.removeChild(node)
      }

      // Remove from registry
      nodes.delete(nodeID)

      return { ok: true }
    } catch (e) {
      return { ok: false, error: e.message }
    }
  }

  function insertAt(parentId, nodeId, index) {
    try {
      const parent = nodes.get(parentId);
      const child = nodes.get(nodeId);

      if (!parent || !child) {
        return { ok: false, error: "insertAt: invalid-node" };
      }

      // clamp index
      const children = parent.childNodes;
      const refNode = (index >= 0 && index < children.length)
        ? children[index]
        : null;

      parent.insertBefore(child, refNode);
      child.__evolve_parent = parentId;

      return { ok: true };
    } catch (e) {
      return { ok: false, error: e.message };
    }
  }


  // Apply props to specific node
  function applyProps(el, props) {
    for (const [k, v] of Object.entries(props)) {
      if (k === "style" && typeof v === "object") {
        Object.assign(el.style, v);
      } else if (k.startsWith("on") && typeof v === "string") {
        // v should be callbackId string from WASM/Python
        // We'll attach an event listener that calls kernel.async.call(callbackId, [eventData])
        const eventName = k.slice(2).toLowerCase();
        const cbId = Number(v);
        el.addEventListener(eventName, (ev) => {
          const eventData = { type: ev.type, targetId: findNodeId(el) };
          asyncCall(cbId, [eventData]);
        });
      } else if (k === "textContent") {
        el.textContent = v;
      } else {
        el.setAttribute(k, v);
      }
    }
  }
  // find the node in existing node registery
  function findNodeId(node) {
     return node.__evolve_id || null
  }
  // update props of the node like style
  function update(nodeId, props = {}) {
    const el = nodes.get(nodeId);
    if (!el) return { ok: false, error: "no-such-node" };
    try {
      applyProps(el, props);
      return { ok: true, value: nodeId };
    } catch (e) {
      return { ok: false, error: e.message };
    }
  }

  // Add children to parent node
  function append(parentId, nodeId) {
    const parent = nodes.get(parentId)
    const child = nodes.get(nodeId)

    if (!parent || !child) return { ok: false, error: "invalid-node" }

    parent.appendChild(child)
    child.__evolve_parent = parentId

    return { ok: true }
  }


  // for injecting evolve in other html web-pages(eg. through extensions)
  function query(selector) {
    const el = document.querySelector(selector);
    if (!el) return { ok: true, value: null };
    const id =
      findNodeId(el) ||
      (function () {
        // register external node
        const nid = genNodeId();
        nodes.set(nid, el);
        el.__evolve_id = nid
        return nid;
      })();
    return { ok: true, value: id };
  }

  

  // Async/callback bridge , to register a callback function
  function registerCallback(fn) {
    const id = genCallbackId();
    callbacks.set(id, fn);
    return { ok: true, value: id };
  }

  // remove callback function when unused
  function unregisterCallback(id) {
    callbacks.delete(Number(id));
    return { ok: true };
  }
  // async calls to get fully completed and predictable response for time-consuming tasks
  async function asyncCall(cbId, args = []) {
    const fn = callbacks.get(Number(cbId));
    if (!fn) {
      console.warn("Kernel: missing callback", cbId);
      return { ok: false, error: "missing-callback" };
    }
    try {
      // Allow callbacks that return promises
      const result = await Promise.resolve(fn(...args));
      return { ok: true, value: result };
    } catch (e) {
      return { ok: false, error: e.message };
    }
  }

  // FS (simple in-memory for prototype) non-persistent
  const fs = new Map();

  function fs_read(path) {
    return { ok: true, value: fs.get(path) ?? null };
  }
  function fs_write(path, contents) {
    fs.set(path, contents);
    return { ok: true };
  }

  // network fetch wrapper
  // any external API request will go through this
  async function net_fetch(url, options = {}) {
    try {
      const res = await fetch(url, options);
      const text = await res.text();
      return {
        ok: true,
        value: {
          status: res.status,
          headers: Object.fromEntries(res.headers.entries()),
          body: text,
        },
      };
    } catch (e) {
      return { ok: false, error: e.message };
    }
  }


  // for accessng window functions
  const location = {
    getPath() {
      return window.location.pathname;
    },

    push(path) {
      if (typeof path !== "string") {
        return { ok: false, error: "path must be string" };
      }
      window.history.pushState({}, "", path);
      return { ok: true };
    },

    replace(path) {
      if (typeof path !== "string") {
        return { ok: false, error: "path must be string" };
      }
      window.history.replaceState({}, "", path);
      return { ok: true };
    },

    onChange(cbId) {
      const id = Number(cbId);

      if (!callbacks.has(id)) {
        return { ok: false, error: "Callback not found" };
      }

      window.addEventListener("popstate", () => {
        const fn = callbacks.get(id);
        if (fn) fn();
      });

      return { ok: true };
    },
    forward() {
      window.history.forward();
      return { ok: true };
    },

    back() {
      window.history.back();
      return { ok: true };
    }

};


  // Public API
  return {
    log,
    dom: { create,remove,insertAt, update, append, query },
    location,
    registerCallback,
    unregisterCallback,
    asyncCall,
    fs: { read: fs_read, write: fs_write },
    net: { fetch: net_fetch },
    _internal: { nodes }, // for debugging in dev only
  };
})();

// attach to window(Global acccess)
window.EvolveKernel = EvolveKernel;
console.log("[Kernel] initialized");
